packages/grafana-data/src/types/logs.ts

```diff
@@ -260,3 +260,37 @@ export const hasLogsContextUiSupport = (datasource: unknown): datasource is Data
 
   return withLogsSupport.getLogRowContextUi !== undefined;
 };
+
+export interface AnalyzeQueryOptions {
+  check: string;
+  attributes: KeyValue<string>;
+}
+
+/**
+ * Data sources that support query manipulation through `modifyQuery` and `analyzeQuery` in Explore.
+ * Allows for interactions such as changing the query from logs details, or displaying filter status.
+ * @internal
+ */
+export interface DataSourceWithQueryManipulationSupport<TQuery extends DataQuery, TAnalyzeQueryResult = boolean> {
+  /**
+   * Used in explore
+   */
+  modifyQuery(query: TQuery, action: QueryFixAction): TQuery;
+
+  /**
+   * Used in explore for Log details
+   *
+   * @alpha
+   */
+  analyzeQuery?(query: TQuery, options: AnalyzeQueryOptions): TAnalyzeQueryResult;
```

Comment:
i wonder if it would make sense to do a performance-optimalization here, and modify this to be able to answer the query-has-filter question for multiple filters, in the same method-call... because, the datasource will probably do this:
1. preprocess the query (for Loki this is a lezer-tree-construction i assume)
2. answer based on the preprocessed data

and, in a logs-detail, we usually need to handle multiple "possible filter"s, so this would mean:
- ask datasource if filter1 is in query
    - datasource builds lezer-tree
    - datasource answers based on lezer-tree
- ask datasource if filter2 is in query
    - datasource builds lezer-tree
    - datasource answers based on lezer-tree
- ask datasource if filter3 is in query
    - datasource builds lezer-tree
    - datasource answers based on lezer-tree

would be good if this could become:
- ask datasource if query contains the following filters: filter1, filter2, filter3
    - datasource builds lezer-tree
    - datasource checks filter1 baed on lezer tree (maybe it could even check for all 3 in a single tree-traversal)
    - datasource checks filter2 baed on lezer tree
    - datasource checks filter3 baed on lezer tree
    - datasource answers

 would be good if the datasource could preprocess only once, and then answer multiple questions.

Reply:
That's a great idea for a follow up. The current implementation works in sync with the original implementation of the 🔍  filters, which add or remove filters from all the queries, when multiple queries or data sources are being used.